controllers/scan.js

```diff
@@ -1,15 +1,68 @@
 "use strict";
 
-const sha1 = require("../sha1-utils");
+const crypto = require("crypto");
+
+const AppConstants = require("../app-constants");
+const { FluentError } = require("../locale-utils");
 const HIBP = require("../hibp");
+const mozlog = require("../log");
+const sha1 = require("../sha1-utils");
+
+
+const log = mozlog("controllers.scan");
+
+
+function _decryptPageToken(encryptedPageToken) {
+  const decipher = crypto.createDecipher("aes-256-cbc", AppConstants.COOKIE_SECRET);
+  const decryptedPageToken = JSON.parse([decipher.update(encryptedPageToken, "hex", "utf8"), decipher.final("utf8")].join(""));
+  return decryptedPageToken;
+}
+
+
+function _validatePageToken(pageToken, req) {
+  const requestIP = req.ip;
+  const pageTokenIP = pageToken.ip;
+  if (pageToken.ip !== requestIP) {
+    log.error("_validatePageToken", {msg: "IP mis-match", pageTokenIP, requestIP});
+    return false;
+  }
+  if (Date.now() - new Date(pageToken.date) >= AppConstants.PAGE_TOKEN_TIMER * 1000) {
+    log.error("_validatePageToken", {msg: "expired pageToken"});
+    return false;
+  }
+  /* TODO: block on scans-per-ip instead of scans-per-timespan
+  if (req.session.scans.length > 5) {
+    console.log("too many scans this session");
+    return res.render("error");
+  }
+  if (!req.session.scans.includes(emailHash)) {
+    console.log(`adding ${emailHash} to session scans`);
+    req.session.scans.push(emailHash);
+  }
+  */
+  return true;
+}
 
 
 async function post (req, res) {
   const emailHash = req.body.emailHash;
+  const encryptedPageToken = req.body.pageToken;
   let featuredBreach = null;
   let userAccountCompromised = false;
   let foundBreaches = [];
 
+  // for #688: use a page token to check for bot scans
+  if (AppConstants.PAGE_TOKEN_TIMER > 0) {
+    if (!encryptedPageToken) {
+      throw new FluentError("error-no-page-token");
+    }
+    const decryptedPageToken = _decryptPageToken(encryptedPageToken);
+    if (!_validatePageToken(decryptedPageToken, req)) {
+      throw new FluentError("error-invalid-page-token");
```

Comment:
I didn't find `"error-no-page-token"` (L57) or `"error-invalid-page-token"` in any of the .ftl files. Not sure if we need to add the strings still since they appear to be user/bot facing.

Reply:
Yes, will need to add some strings for this. But I don't know if it needs @betsymi to write them, or if I can just write them myself. @betsymi - do you want to write error messages here?

Reply:
Yes, happy to. I'm having a hard time following what the errors are themselves though. @groovecoder could you help me out by drafting what these are supposed to communicate, and I can work on making them more user-friendly?

Reply:
@betsymi @flodolo right now I just have a single error string: "Your network may be performing too many scans. Please try again later."

Is that good?

Reply:
I wonder if that's a bit confusing for normal users? (i.e. what's "my network"?) Not sure what's the target and frequency of this message showing up though.

Maybe something like: "We detected too many scans originating from your address and the request has been blocked. Please try again later."

Reply:
What makes this error appear? Is it that a bot is trying to scan too many email addresses in quick succession?

Reply:
So far yes - this is aimed at bots. The normal users who are likely to encounter this would be people who are using "Scan another email" box repeatedly.

I agree "network" may not be the right term. But "address" might feel weird too - will people think we're somehow looking at their physical address? Should we be afraid of using the term "IP Address"? Google and many others use it all the time when they don't want me using their services from Tor.

(But as a scraper-writing colleague of mine pointed out, we shouldn't be anti-bot 😜 ... i.e., we don't really want humans to be scanning  dozens of email addresses either.)

Reply:
You tried to search too many emails in a short time period. For security reasons, this IP address is temporarily blocked from new searches. You can try again later.